ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -431,52 +433,57 @@ private PlanNode buildSourceForJoin(
   }
 
   private PlanNode buildSourceNode() {
-
-    final List<AliasedDataSource> sources = analysis.getAllDataSources();
-
     if (!analysis.isJoin()) {
-      return buildNonJoinNode(sources);
+      return buildNonJoinNode(analysis.getFrom());
     }
 
-    if (sources.size() == 1) {
-      throw new IllegalStateException("Expected more than one source. Got " + sources.size());
-    } else if (sources.size() != 2) {
-      throw new KsqlException(
-          "Invalid join criteria specified; KSQL does not support multi-way joins.");
+    final List<JoinInfo> joinInfo = analysis.getOriginal().getJoin();
+    final JoinTree.Node tree = JoinTree.build(joinInfo);
+    if (tree instanceof JoinTree.Leaf) {
+      throw new IllegalStateException("Expected more than one source:"
+          + analysis.getAllDataSources());
     }
 
-    final AliasedDataSource left = sources.get(0);
-    final AliasedDataSource right = sources.get(1);
+    return buildJoin((Join) tree);
+  }
 
-    final List<JoinInfo> joinInfo = analysis.getOriginal().getJoin();
 
-    final PlanNode leftSourceNode = buildSourceForJoin(
-        left,
-        "Left",
-        joinInfo.get(0).getLeftJoinExpression()
-    );
+  /**
+   * @param root  the root of the Join Tree
+   * @return the PlanNode representing this Join Tree
+   */
+  private PlanNode buildJoin(final Join root) {
+    final PlanNode left;
+    if (root.getLeft() instanceof JoinTree.Join) {
+      throw new KsqlException(
+          "Invalid join criteria specified; KSQL does not support multi-way joins.");
```

Comment:
in the next PR, we will recurse here instead of throwing an exception
```suggestion
      left = buildJoin(root.getLeft());
```